src/lib/Simplify.hs

```diff
@@ -162,8 +163,11 @@ simplifyOp op = case op of
 simplifyHof :: Hof -> SimplifyM Atom
 simplifyHof hof = case hof of
   For d lam -> do
-    ~(lam', Nothing) <- simplifyLam lam
-    emit $ Hof $ For d lam'
+    ~(lam'@(Lam (Abs i _)), recon) <- simplifyLam lam
+    ans <- emit $ Hof $ For d lam'
+    case recon of
+      Nothing -> return ans
+      Just f  -> buildLam i TabArrow $ \i' -> app ans i' >>= f
```

Comment:
When would that happen? I thought that `for`s always have to have a data return value

Reply:
Good question! This is for linearization. Linearization produces these sorts of `fors`. For example, we might linearize this:

```
for i.
  x = ...
  y = ...
  y
```

To this:

```
for i.
  x = ...
  y = ...
  (y, \t1 t2 ... .
          xt = ...
          yt = ... x ...
          y)
```

This way, AD doesn't have to worry about "capturing" the intermediate values like `x`. It just closes over them, and lets the general defunctionalization machinery take care of them.

Reply:
Right! It would be interesting to consider an extension to tables that would allow you to store functions inside them, as long as all the functions have the same code and only differ in the values they close over. That seems easy enough to handle, especially now that we use table lambdas that could load the respective data and return a lambda atom!